### ISSUE TITLE
testpinyindictionary segfault with boost 1.81.0

### ISSUE THREAD

**Issue by sgn** [fcitx/libime#40]
The segfault caused by nullptr-dereference in line 381

https://github.com/fcitx/libime/blob/e199ce44fd85b86e0a906819fa93d4fadda3a2ca/src/libime/pinyin/pinyindictionary.cpp#L376-L381

libime finds the path by `Compatible{Key,Hash,Predicate}`:
https://github.com/fcitx/libime/blob/e199ce44fd85b86e0a906819fa93d4fadda3a2ca/src/libime/core/lrucache.h#L75-L80
but the `insert` check by `key_type`:
https://github.com/fcitx/libime/blob/e199ce44fd85b86e0a906819fa93d4fadda3a2ca/src/libime/core/lrucache.h#L43-L44

**Comment by sgn** [comment]
This diff fixes the segfault but not sure if it's your intend
```diff
Index: libime-1.0.15/src/libime/pinyin/pinyindictionary.cpp
===================================================================
--- libime-1.0.15.orig/src/libime/pinyin/pinyindictionary.cpp
+++ libime-1.0.15/src/libime/pinyin/pinyindictionary.cpp
@@ -376,6 +376,9 @@ bool PinyinDictionaryPrivate::matchWords
         auto *result =
             matchCache.find(path.path_, context.hasher_, context.hasher_);
         if (!result) {
+            result = matchCache.find(context.hasher_.pathToPinyins(path.path_));
+        }
+        if (!result) {
             result =
                 matchCache.insert(context.hasher_.pathToPinyins(path.path_));
             result->clear();
```

**Comment by wengxt** [comment]
This is called heterogeneous lookup. You can do the look up without create an actual key type. 

I think the key issue is that boost changes the way it hashes std::string, that breaks the assumption of hasher.

**Comment by sgn** [comment]
Their release notes seems to indicate there is a major change in Boost.Unordered.

**Comment by wengxt** [comment]
I'd suggest it's about boost::hash<std::string> specifically.

**Comment by wengxt** [comment]
I can see there's a major update of "container hash" in 1.81. I guess that changes how hash against std::string works.

The fix should be use our own string hash implementation to we won't be break by upstream change in the future.

@sgn  can you check if master fixes the issue?

**Comment by sgn** [comment]
Yes, the change fixed libime for boost 1.81.0
